src/lib/Inference.hs

```diff
@@ -1361,16 +1393,18 @@ introDictTys (h:t) m = buildPiInf (getNameHint @String "_autoq") ClassArrow h \_
   ListE t' <- sinkM $ ListE t
   (Pure,) . Pi <$> (introDictTys t' m)
 
-checkMethodDef :: (EmitsBoth o, Inferer m)
-               => ClassName o -> [Type o] -> UMethodDef i -> m i o (Int, Atom o)
+checkMethodDef :: (EmitsInf o, Inferer m)
+               => ClassName o -> [MethodType o] -> UMethodDef i -> m i o (Int, Block o)
 checkMethodDef className methodTys (UMethodDef ~(InternalName sourceName v) rhs) = do
   MethodBinding className' i _ <- substM v >>= lookupEnv
   when (className /= className') do
-    ClassBinding (ClassDef classSourceName _ _) _ <- lookupEnv className
+    ClassBinding (ClassDef classSourceName _ _ _ _) <- lookupEnv className
     throw TypeErr $ pprint sourceName ++ " is not a method of "
                  ++ pprint classSourceName
-  let methodTy = methodTys !! i
-  rhs' <- checkSigma rhs methodTy
+  let MethodType _ methodTy = methodTys !! i
+  rhs' <- buildBlockInf do
```

Comment:
It's a real change! Notice that `checkMethodDef` now just uses `EmitsInf o` instead of `EmitsBoth o`. Methods are now blocks, and the decls get evaluated at projection time instead of at instance definition time and dictionary synthesis time. (Similar to the "automatic thunking" we had discussed.) So defining an instance doesn't actually emit any decls anymore.